proposals/0000-decompose-newtypes.rst

```diff
@@ -0,0 +1,192 @@
+``[W] Int ~R Bool``, and complain that it can't prove that::
+
+    Foo.hs:12:7: error: [GHC-18872]
+        • Couldn't match type ‘Int’ with ‘Bool’
+            arising from a use of ‘coerce’
+        • In the expression: coerce x
+          In an equation for ‘f’: f x = coerce x
+       |
+    12 | f x = coerce x
+       |       ^^^^^^
+
+But it's totally wrong to claim that
+we need ``Int ~R Bool``, because of the type family.  (E.g. suppose ``F Int = Char`` and ``F Bool = Char``.)
+
+So the error message is bogus.  In more complicated situations we might unify type variables
+that should not be unified, causing more misleading, far-away errors.
+
+In short,
+
+* the decomposition step for newtypes is always *sound* (if typechecking succeeds, all is well),
+* but it is not always *complete* (typechecking may fail when it shouldn't).
+
+So you may say "do not decompose newtype wanteds".  But then suppose
+we are trying to prove ``[W] (IO Int) ~R (IO Age)``, where::
+
+   newtype Age = MkAge Int
+   newtype IO a = MkIO (State -> (State,a))
+
+and ``MkAge`` is in scope, but ``MkIO`` is not (being rightly abstract).
+If we do not decompose newtypes, GHC would be totally stuck, which is not good for the programmer.
+
+So GHC has a weird special case, where it says "as a last resort, decompose the newtype unless there are
+in-scope Given equalities".  The "unless" part is an ad-hoc heuristic,
+intended to make incompleteness harder to trigger.
+(Interested parties can look at
+``can_decompose`` in ``GHC.Tc.Solver.Canonical.canTyConApp``, and the
+discussion on `this merge request <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9282>`_.
+
+But even with the horrid special case, we still get the bogus ``Int ~ Bool`` error message.
+
+
+The CO-NTH rule
+:::::::::::::::
+
+For a newtype ``T``, given ``[W] (T t1) ~R (T t2)``
+can we ever decompose to ``[W] t1 ~R t2``, without losing completeness?
+
+The relevant paper is `Safe zero-cost coercions for Haskell (journal version) <https://simon.peytonjones.org/safe-coercions-2016/>`_.
+Here's the CO-NTH rule in the paper, Fig 5::
+
+  G |- g : H taus ~R H sigmas    rs is a prefix of roles(H)     H is not a newtype
+  ---------------------------------------------------------------------------------------    CO-NTH
+     G |- Nth(i) G : taus(i) ~rs(i) sigmas(i)
+
+The CO-NTH rule governs completeness of decomposition for wanted constraints: if CO-NTH holds
+for each argument i, it means that nothing is lost by decomposing.
+
+But note "H is not a newtype"; that means **we can never decompose a newtype** except
+at risk of incompleteness.
+But in the case of the newtype ``IO`` above, it is easy to see that
+if we know ``(IO t1) ~R (IO t2)`` then we certainly know ``t1 ~R t2``, and vice versa.
+**Nothing is lost by decomposing ``IO``!**  CO-NTH is too conservative.
+
+The proposal below (see Proposed Change Specification) strengthens CO-NTH to allow
+this decomposition.
+
+
```

Comment:
I am unmoved by the current motivation, because I don't really see the problem today.

Instead, I think a different motivation would hold more water: You want `[G] (IO Int) ~R (IO Age)` to imply `[G] Int ~R Age`, which indeed we do not have today. But actually I find the usefulness of that implication to be minimal, so I'm still stuck on the motivation piece.